File: isegoria/src/main/java/com/eulersbridge/isegoria/auth/signup/SignUpFragment.java
package com.eulersbridge.isegoria.auth.signup;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ScrollView;
import android.widget.Spinner;

import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.auth.AuthViewModel;
import com.eulersbridge.isegoria.network.api.models.Country;
import com.eulersbridge.isegoria.network.api.models.Institution;
import com.eulersbridge.isegoria.util.Utils;
import com.eulersbridge.isegoria.util.data.SimpleTextWatcher;

import java.util.Calendar;
import java.util.List;

public class SignUpFragment extends Fragment {

    private Button signUpButton;
    private ImageView backButton;

    private EditText givenNameField;
    private EditText familyNameField;
    private EditText emailField;
    private EditText newPasswordField;
    private EditText confirmNewPasswordField;
    private EditText genderField;

    private Spinner countrySpinner;
    private Spinner institutionSpinner;
    private Spinner yearOfBirthSpinner;

	private ArrayAdapter<Country> countryAdapter;
	private ArrayAdapter<Institution> institutionAdapter;

	private SignUpViewModel viewModel;
	private AuthViewModel authViewModel;

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View rootView = inflater.inflate(R.layout.sign_up_fragment, container, false);

		if (getActivity() != null)
		    authViewModel = ViewModelProviders.of(getActivity()).get(AuthViewModel.class);

		viewModel = ViewModelProviders.of(this).get(SignUpViewModel.class);

        final ScrollView scrollContainer = rootView.findViewById(R.id.sign_up_container);

        boolean hasTranslucentStatusBar = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
        if (hasTranslucentStatusBar) {
            int additionalTopPadding = Math.round(22 * getResources().getDisplayMetrics().density);

            scrollContainer.setPadding(
                    scrollContainer.getPaddingLeft(),
                    scrollContainer.getPaddingTop() + additionalTopPadding,
                    scrollContainer.getPaddingRight(),
                    scrollContainer.getPaddingBottom());
        }

        backButton = rootView.findViewById(R.id.sign_up_back_button);

        signUpButton = rootView.findViewById(R.id.sign_up_next_button);

        givenNameField = rootView.findViewById(R.id.sign_up_given_name);
        familyNameField = rootView.findViewById(R.id.sign_up_family_name);
        emailField = rootView.findViewById(R.id.sign_up_email);
        newPasswordField = rootView.findViewById(R.id.sign_up_new_password);
        confirmNewPasswordField = rootView.findViewById(R.id.sign_up_confirm_new_password);
        genderField = rootView.findViewById(R.id.sign_up_gender);

        AppCompatActivity activity = (AppCompatActivity)getActivity();

        countrySpinner = rootView.findViewById(R.id.sign_up_country);

        countryAdapter = new ArrayAdapter<>(activity, android.R.layout.simple_spinner_item);
        countryAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        countrySpinner.setAdapter(countryAdapter);

        institutionSpinner = rootView.findViewById(R.id.sign_up_institution);
        institutionAdapter = new ArrayAdapter<>(activity, android.R.layout.simple_spinner_item);
        institutionAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        institutionSpinner.setAdapter(institutionAdapter);
        
        yearOfBirthSpinner = rootView.findViewById(R.id.sign_up_birth_year);
		ArrayAdapter<String> spinnerYearOfBirthArrayAdapter = new ArrayAdapter<>(activity, android.R.layout.simple_spinner_item);

		final int year = Calendar.getInstance().get(Calendar.YEAR);

        // Allow users to be in the age range 12 to 100
        for (int i = (year - 100); i <= (year - 12); i++) {
        	spinnerYearOfBirthArrayAdapter.add(String.valueOf(i));

            if (i == 1990) yearOfBirthSpinner.setSelection(i);
        }
        spinnerYearOfBirthArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        yearOfBirthSpinner.setAdapter(spinnerYearOfBirthArrayAdapter);

        setupViewListeners();

        viewModel.getCountries().observe(this, countries -> {
            if (countries != null)
                setCountries(countries);
        });

        Utils.showKeyboard(activity.getWindow());
		
		return rootView;
	}

    private void setupViewListeners() {
        backButton.setOnClickListener(view -> {
            if (getActivity() != null) {
                authViewModel.onSignUpBackPressed();
                getActivity().onBackPressed();
            }
        });

        signUpButton.setOnClickListener(view -> {
            backButton.setEnabled(false);
            signUpButton.setEnabled(false);

            SignUpUser newUser = viewModel.getSignUpUser();
            if (newUser == null) {
                backButton.setEnabled(true);
                signUpButton.setEnabled(true);

            } else {
                authViewModel.signUpUser.setValue(newUser);
            }
        });

        givenNameField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setGivenName(value.toString())
        ));
        familyNameField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setFamilyName(value.toString())
        ));
        emailField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setEmail(value.toString())
        ));
        newPasswordField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setPassword(value.toString())
        ));
        confirmNewPasswordField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setConfirmPassword(value.toString())
        ));
        genderField.addTextChangedListener(new SimpleTextWatcher(value ->
            viewModel.setGender(value.toString())
        ));

        countrySpinner.setOnItemSelectedListener(new SimpleOnItemSelectedListener() {
            @Override
            void onItemSelected(int position) {
                institutionSpinner.setEnabled(false);

                List<Institution> institutions = viewModel.onCountrySelected(position);
                if (institutions != null) {
                    institutionAdapter.clear();
                    institutionAdapter.addAll(institutions);
                }

                institutionSpinner.setEnabled(true);
            }
        });

        institutionSpinner.setOnItemSelectedListener(new SimpleOnItemSelectedListener() {
            @Override
            void onItemSelected(int position) {
                viewModel.onInstitutionSelected(position);
            }
        });

        yearOfBirthSpinner.setOnItemSelectedListener(new SimpleOnItemSelectedListener() {
            @Override
            void onItemSelected(int position) {
                String birthYear = (String) yearOfBirthSpinner.getSelectedItem();
                viewModel.onBirthYearSelected(birthYear);
            }
        });
    }

    private void setCountries(List<Country> newCountries) {
        countrySpinner.setEnabled(false);

        if (newCountries != null) {
            countryAdapter.clear();
            countryAdapter.addAll(newCountries);
        }

        countrySpinner.setEnabled(true);
	}
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/auth/signup/SignUpFragment.kt
package com.eulersbridge.isegoria.auth.signup

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.core.view.updatePadding
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.auth.AuthViewModel
import com.eulersbridge.isegoria.auth.onTextChanged
import com.eulersbridge.isegoria.network.api.model.Country
import com.eulersbridge.isegoria.network.api.model.Institution
import com.eulersbridge.isegoria.util.extension.HintAdapter
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.onItemSelected
import com.eulersbridge.isegoria.util.extension.setKeyboardVisible
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.sign_up_fragment.*
import java.util.*
import javax.inject.Inject

class SignUpFragment : Fragment() {

    @Inject
    internal lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: SignUpViewModel

    final var gender_hint = "Gender"
    final var yearOfBirth_hint = "Year of birth"
    final var institution_hint = "Institution"

    private val authViewModel: AuthViewModel by lazy {
        ViewModelProviders.of(requireActivity())[AuthViewModel::class.java]
    }

    private lateinit var countryAdapter: ArrayAdapter<Country>
    private lateinit var institutionAdapter: ArrayAdapter<Institution>

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        viewModel = ViewModelProviders.of(this, modelFactory)[SignUpViewModel::class.java]
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
            = inflater.inflate(R.layout.sign_up_fragment, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val hasTranslucentStatusBar = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
        if (hasTranslucentStatusBar) {
            val additionalTopPadding = Math.round(22 * resources.displayMetrics.density)
            container.updatePadding(top = container.paddingTop + additionalTopPadding)
        }

//        val plants = arrayOf("Select an item...", "California sycamore", "Mountain mahogany", "Butterfly weed", "Carrot weed")
//        val plantsList = ArrayList(Arrays.asList(plants))
        val genderAdapter = HintAdapter<String>(view.context, android.R.layout.simple_spinner_item)
        genderAdapter.add("Male")
        genderAdapter.add("Female")
        genderAdapter.add("Others")
        genderAdapter.add(gender_hint)
        genderAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        genderSpinner.adapter = genderAdapter
        genderSpinner.setSelection(genderAdapter.count)


        countryAdapter = ArrayAdapter(view.context, android.R.layout.simple_spinner_item)
        countryAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        countrySpinner.apply {
            isEnabled = false
            adapter = countryAdapter
        }

        institutionAdapter = ArrayAdapter(activity, android.R.layout.simple_spinner_item)
        institutionAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        loginInstitutionSpinner.adapter = institutionAdapter

        val spinnerYearOfBirthArrayAdapter = HintAdapter<String>(view.context, android.R.layout.simple_spinner_item)

        val year = Calendar.getInstance()[Calendar.YEAR]

        // Allow users to be in the age range 12 to 100
        for (i in year - 100..year - 12) {
            spinnerYearOfBirthArrayAdapter.add(i.toString())

            if (i == 1990)
                birthYearSpinner.setSelection(i)
        }
        spinnerYearOfBirthArrayAdapter.add(yearOfBirth_hint)
        spinnerYearOfBirthArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        birthYearSpinner.adapter = spinnerYearOfBirthArrayAdapter
        birthYearSpinner.setSelection(spinnerYearOfBirthArrayAdapter.count)
        createViewListeners()

        observe(viewModel.countries) {
            setCountries(it!!)
        }

        activity?.setKeyboardVisible(true)
    }

    private fun createViewListeners() {
        backButton.setOnClickListener {
            activity?.let {
                authViewModel.onSignUpBackPressed()
                it.onBackPressed()
            }
        }

        nextButton.setOnClickListener {
            it.isEnabled = false
            backButton.isEnabled = false

            val newUser = viewModel.getSignUpUser()
            if (newUser == null) {
                backButton.isEnabled = true
                it.isEnabled = true

            } else {
                authViewModel.signUpUser.value = newUser
            }
        }

        givenName.onTextChanged { if (it != null) viewModel.setGivenName(it) }
        familyName.onTextChanged { if (it != null) viewModel.setFamilyName(it) }
        email.onTextChanged { if (it != null) viewModel.setEmail(it) }
        newPassword.onTextChanged { if (it != null) viewModel.setPassword(it) }
        confirmNewPassword.onTextChanged { if (it != null) viewModel.setConfirmPassword(it) }
//        genderSpinner.onTextChanged { if (it != null) viewModel.setGender(it) }
        genderSpinner.onItemSelected {
            val gender = genderSpinner.selectedItem as String
            if (gender != gender_hint)
                viewModel.setGender(gender)
        }

        countrySpinner.onItemSelected { position -> updateInstitutionSpinner(position) }

        loginInstitutionSpinner.onItemSelected { position -> viewModel.onInstitutionSelected(position) }

        birthYearSpinner.onItemSelected {
            val birthYear = birthYearSpinner.selectedItem as String
            if (birthYear != yearOfBirth_hint)
                viewModel.setBirthYear(birthYear)
        }
    }

    private fun updateInstitutionSpinner(position: Int) {
        loginInstitutionSpinner.isEnabled = false

        viewModel.onCountrySelected(position)?.let { institutions ->
            institutionAdapter.clear()
            institutionAdapter.addAll(institutions)
        }

        loginInstitutionSpinner.isEnabled = true
    }

    private fun setCountries(newCountries: List<Country>) {
        countryAdapter.apply {
            clear()
            addAll(newCountries)
        }

        updateInstitutionSpinner(0)

        countrySpinner.isEnabled = true
    }
}
